core/lib/zksync_core/src/sync_layer/gossip/storage/mod.rs

```diff
@@ -0,0 +1,225 @@
+        else {
+            return Ok(None);
+        };
+        let block = sync_block_to_consensus_block(block)?;
+        Ok(Some(block))
+    }
+
+    async fn first_block(&self) -> anyhow::Result<FinalBlock> {
+        let mut storage = self.storage().await?;
+        Self::block(&mut storage, MiniblockNumber(0))
+            .await?
+            .context("Genesis miniblock not present in Postgres")
+    }
+
+    async fn sealed_miniblock_number(&self) -> anyhow::Result<BlockNumber> {
+        let mut storage = self.storage().await?;
+        let number = storage
+            .blocks_dal()
+            .get_sealed_miniblock_number()
+            .await
+            .context("Failed getting sealed miniblock number")?;
+        Ok(BlockNumber(number.0.into()))
+    }
+}
+
+#[async_trait]
+impl BlockStore for PostgresBlockStorage {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        ctx.wait(self.head_block())
+            .await?
+            .map_err(StorageError::Database)
+    }
+
+    async fn first_block(&self, ctx: &ctx::Ctx) -> StorageResult<FinalBlock> {
+        ctx.wait(self.first_block())
+            .await?
+            .map_err(StorageError::Database)
+    }
+
+    async fn last_contiguous_block_number(&self, ctx: &ctx::Ctx) -> StorageResult<BlockNumber> {
+        ctx.wait(self.sealed_miniblock_number())
+            .await?
+            .map_err(StorageError::Database)
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: BlockNumber,
+    ) -> StorageResult<Option<FinalBlock>> {
+        ctx.wait(async {
+            let number = u32::try_from(number.0).context("block number is too large")?;
+            let mut storage = self.storage().await?;
+            Self::block(&mut storage, MiniblockNumber(number)).await
+        })
+        .await?
+        .map_err(StorageError::Database)
+    }
+
+    async fn missing_block_numbers(
+        &self,
+        _ctx: &ctx::Ctx,
+        _range: ops::Range<BlockNumber>,
+    ) -> StorageResult<Vec<BlockNumber>> {
+        Ok(vec![]) // The storage never has missing blocks by construction
+    }
+
+    fn subscribe_to_block_writes(&self) -> watch::Receiver<BlockNumber> {
+        self.block_sender.subscribe()
+    }
+}
+
+#[async_trait]
+impl ContiguousBlockStore for PostgresBlockStorage {
+    async fn schedule_next_block(&self, ctx: &ctx::Ctx, block: &FinalBlock) -> StorageResult<()> {
+        // last_in_batch` is always set to `false` by this call; it is properly set by `CursorWithCachedBlock`.
+        let fetched_block =
+            FetchedBlock::from_gossip_block(block, false).map_err(StorageError::Database)?;
+        let actions = sync::lock(ctx, &self.cursor).await?.advance(fetched_block);
+        ctx.wait(async {
+            for actions_chunk in actions {
+                self.actions.push_actions(actions_chunk).await;
```

Comment:
is push_actions cancel safe? Is executing this loop partially cancel safe?

Reply:
Executing a loop partially should be cancel-safe (each `actions_chunk` corresponds to a separate L2 block, so it is OK if a further chunk is lost). There is a caveat: it ceases to be cancel-safe if we assume that the surrounding struct (`PostgresBlockStorage`) can be used after cancellation – which we currently don't and likely don't ever plan to, but that's non-local reasoning. If we assume this possibility, than executing a loop partially is **not** cancel-safe since it'll lead to a mismatch between the state of the action queue and the cursor (i.e., `actions` and `cursor` fields).

`push_actions` is somewhat similar – if we assume that the action queue consumer stops together with cancellation and doesn't resume, we should be fine. In the general case, losing some actions in the queue will likely lead to panics or maybe even to the corrupted node state (not sure whether it's possible to trigger).

Given all this, would you advise not to use `ctx.wait()` here?

Reply:
Yes, I'd suggest to await unconditionally if cancellation would break the state, as long as push_actions is known to eventually complete.

Reply:
otherwise annotate the whole schedule_next_block method that it invalidates the PostgresBlockStorage object if cancelled.